tests/test_reordering.f90

```diff
@@ -0,0 +1,225 @@
+            end do
+        end do
+    end do
+    if (.not. pass_X) print *, "Error in X direction for index reversing"
+    if (.not. pass_Y) print *, "Error in Y direction for index reversing"
+    if (.not. pass_Z) print *, "Error in Z direction for index reversing"
+
+    allpass = (pass_X .and. pass_Y .and. pass_Z)
+ 
+
+    ! Test reordering
+    u_x => allocator%get_block(DIR_X)
+    u_y => allocator%get_block(DIR_Y)
+    u_z => allocator%get_block(DIR_Z)
+    u_x_original => allocator%get_block(DIR_X)
+
+    dims(:) = allocator%xdims_padded(:)
+    allocate (u_array(dims(1), dims(2), dims(3)))
+
+    call random_number(u_array)
+
+#ifdef CUDA
+    allocate (temp_1(dims(1), dims(2), dims(3)))
+    allocate (temp_2(dims(1), dims(2), dims(3)))
+
+    select type (u_x_original)
+    type is (cuda_field_t)
+       u_x_original%data_d = u_array
+    end select
+#else
+    select type (u_x_original)
+    type is (field_t)
+       u_x_original%data = u_array
+    end select
+#endif
+
+    call backend%reorder(u_y, u_x_original, RDR_X2Y)
+    call backend%reorder(u_x, u_y, RDR_Y2X)
+    call check_reorder(allpass, u_x, u_x_original, "X2Y, Y2X")
+
+    call backend%reorder(u_z, u_x, RDR_X2Z)
+    call backend%reorder(u_x, u_z, RDR_Z2X)
+    call check_reorder(allpass, u_x, u_x_original, "X2Z, Z2X")
+
+    call backend%reorder(u_z, u_y, RDR_Y2Z)
+    call backend%reorder(u_x, u_z, RDR_Z2X)
+    call check_reorder(allpass, u_x, u_x_original, "Y2Z, Z2X")
+
+    call backend%reorder(u_y, u_z, RDR_Z2Y)
+    call backend%reorder(u_x, u_y, RDR_Y2X)
+    call check_reorder(allpass, u_x, u_x_original, "Z2Y, Y2X")
+
+    if (allpass) then
+        if (nrank == 0) write(stderr, '(a)') 'ALL TESTS PASSED SUCCESSFULLY.'
+    else
+        error stop 'SOME TESTS FAILED.'
+    end if
+
+    call MPI_Finalize(ierr)
+
+    contains
+
+    subroutine test_index_reversing(pass, i, j, k, dir, SZ, nx, ny, nz)
+        logical, intent(inout) :: pass
+        integer, intent(in) :: i, j, k    ! original indices in the cartesian space
+        integer, intent(in) :: dir
+        integer, intent(in) :: SZ, nx, ny, nz
+        integer :: dir_i, dir_j, dir_k    ! indices in the applicatin storage direction
+        integer :: cart_i, cart_j, cart_k ! newly computed indices in the cartesian space
+
+        call get_index_dir(dir_i, dir_j, dir_k, i, j, k, dir, SZ, nx, ny, nz)
+        call get_index_ijk(cart_i, cart_j, cart_k, dir_i, dir_j, dir_k, dir, SZ, nx, ny, nz)
+
+        if (i /= cart_i .or. j /= cart_j .or. k /= cart_k) then
+            pass = .false.
+        end if
+
+    end subroutine
+
+    subroutine check_reorder(allpass, a, b, message)
+        logical, intent(inout) :: allpass
+        class(field_t), intent(in) :: a, b
+        character(len=*), intent(in) :: message
+        real(dp) :: tol = 1d-8
+
+#ifdef CUDA
+        select type (a); type is (cuda_field_t); temp_1 = a%data_d; end select
```

Comment:
@semi-h Just a thought not related to this issue but would it make sense to have temp_1 stored inside the a structure. We could call something like a%copy_to_host(), that would run this line and we should be able to use a%data just like on the omp backend?
Same we could have a%copy_to_device() to do the opposite.
We would still need the ifdef but it will simplify/clarify code quite a bit

Reply:
For doing that I think we would need to allocate the `%data` component, which currently we keep as a private member to discourage people accessing directly this component. Maybe we can use the `get_field` in the backends for making this cleaner, but I think I'll be making some changes in `get_field` with the cartesian / specialist data structure mapping so it might be better leaving this after `get/set_field` functions are more stable.